src/deterministic_randomizer.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2022 The Tari Project
+// SPDX-License-Identifier: BSD-3-Clause
+
+//! A generic type for deterministic pseudorandom operations: selection and in-place shuffle. We want to ensure the
+//! process is deterministic and portable and forward compatible, so we don't want to rely on specific random selection
+//! behavior. For this reason, we implement the guts of the operations ourselves, and rely only on a deterministic
+//! seeded CSPRNG with trait-enforced security properties.
+
+use std::convert::TryFrom;
+
+use rand_core::{CryptoRng, RngCore, SeedableRng};
+use zeroize::Zeroize;
+
+/// Error representing a failed shuffle
+#[derive(Debug)]
+pub struct RandomizerError;
+
+/// A deterministic randomizer on generic types
+#[derive(Zeroize)]
+pub struct DeterministicRandomizer<R>
+where R: SeedableRng
+{
+    prng: R,
+    seed: <R as SeedableRng>::Seed,
+}
+
+impl<R> DeterministicRandomizer<R>
+where
+    R: CryptoRng + RngCore + SeedableRng,
+    <R as SeedableRng>::Seed: Clone,
+{
+    /// Initialize the randomizer with a seed
+    pub fn new(seed: <R as SeedableRng>::Seed) -> Self {
+        Self {
+            prng: R::from_seed(seed.clone()),
+            seed,
+        }
+    }
+
+    /// Reseed the randomizer with a new seed
+    pub fn reseed(&mut self, seed: <R as SeedableRng>::Seed) {
+        self.prng = R::from_seed(seed.clone());
+        self.seed = seed;
+    }
+
+    /// Reset the randomizer to its original seed
+    pub fn reset(&mut self) {
+        self.reseed(self.seed.clone());
+    }
+
+    /// Shuffle an array in place
+    pub fn shuffle<T>(&mut self, data: &mut [T]) -> Result<(), RandomizerError> {
+        // Restrict the array size to avoid bias
+        let checked_length = u32::try_from(data.len()).map_err(|_| RandomizerError)?;
```

Comment:
Why `u32` here?

Reply:
One reason was that the original draft didn't support `u64` because of underlying PRNG API limits, which isn't the case anymore. The other reason was an assumption that the library should support 32-bit architectures, which (after discussion) is not a requirement.

Changed to assume 64-bit indexing.